src/Workspaces/CoreTest/SolutionTests/ProjectDependencyGraphTests.cs

```diff
@@ -68,45 +98,254 @@ private void VerifyDependencySets(string projectReferences, string expectedResul
         #region GetProjectsThatThisProjectTransitivelyDependsOn
 
         [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
-        public void ProjectDependencyGraph_GetProjectsThatThisProjectTransitivelyDependsOn()
+        public void TestGetProjectsThatThisProjectTransitivelyDependsOn()
+        {
+            VerifyTransitiveReferences(CreateSolutionFromReferenceMap("A"), "A", new string[] { });
+            VerifyTransitiveReferences(CreateSolutionFromReferenceMap("B:A A"), "B", new string[] { "A" });
+            VerifyTransitiveReferences(CreateSolutionFromReferenceMap("C:B B:A A"), "C", new string[] { "B", "A" });
+            VerifyTransitiveReferences(CreateSolutionFromReferenceMap("C:B B:A A"), "A", new string[] { });
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void TestGetProjectsThatThisProjectTransitivelyDependsOnThrowsArgumentNull()
+        {
+            var solution = CreateSolutionFromReferenceMap("");
+
+            Assert.Throws<ArgumentNullException>("projectId",
+                () => solution.GetProjectDependencyGraph().GetProjectsThatThisProjectDirectlyDependsOn(null));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void TestTransitiveReferencesIncrementalUpdateInMiddle()
+        {
+            // We are going to create a solution with the references:
+            //
+            // A -> B -> C -> D
```

Comment:
Should there be a test where the newly updated transitive references can *only* be constructed from existing *transitive* references? That is, something like `A -> B -> C     D -> E -> F`, connect `C -> D`, and make sure A & F get connected. Different from this test because in this test D is in C's direct references and A is in B's direct reverse references.

Reply:
Can we also consider tests with more references being added at the same time (sort of related to Cyrus's Contains/Overlaps comment where the forward & reverse directions don't have _identical_ code because we're always considering adding many forward references at the same time)? I'd be happy to help draw complicated graphs for this purpose.... 😄

Reply:
I, too, would appreciate a complicated drawing of a porpoise.   Perhaps something like this: 

![image](https://user-images.githubusercontent.com/4564579/44599979-c8261880-a78c-11e8-8450-044969a2cde1.png)

Reply:
@dpoeschl I've added a test for your scenario as well. I'll hold on porpoises for now. (Or dolphins, which I think your image there looks like.)